### PR TITLE
Replace firmware implementation of HCSR04

### LINKED ISSUE #208: Port NONSTANDARD_IO functions
- [ ] HCSR04

### PR DESCRIPTION
**Pull request by bessman**
This is basically the same as what I posted [earlier](https://github.com/fossasia/pslab-firmware/pull/118#issuecomment-1076290022) with some minor cleanup and added error checking.

### CODE REVIEW

File: pslab/external/hcsr04.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+    Sensors must have separate trigger and echo pins. First a 10 µs pulse is
+    output on the trigger pin. The trigger pin must be connected to the TRIG
+    pin on the sensor prior to use.
+
+    Upon receiving this pulse, the sensor emits a sequence of sound pulses, and
+    the logic level of its echo pin is also set high.  The logic level goes LOW
+    when the sound packet returns to the sensor, or when a timeout occurs.
+    Timeout occurs if no echo is received within the time slot determinded by
+    the sensor's maximum range.
```
Comment (bessman) [inline pslab/external/hcsr04.py:23]: @CloudyPadmal Please check if this is correct. The datasheet is unclear and different online sources contradict each other on this point.

File: pslab/external/hcsr04.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+        ----------
+        average : int, optional
+            Number of times to repeat the measurement and average the results.
+            Defaults to 10.
+        speed_of_sound : float, optional
+            Speed of sound in air. Defaults to 340 m/s.
+        timeout_distance : float, optional
+            Maximum distance to object. Defaults to 4 m. Measurement times out
+            if the echo is not received within:
+                timeout = timeout_distance / speed_of_sound * 2 seconds
```
Comment (bessman) [inline pslab/external/hcsr04.py:88]: Please check if this is a reasonable default.

**Comment by bessman** [comment]
One drawback of this implementation is that the timeout isn't configurable. The logic analyzer timeout cannot be shorter than the trigger pulse period (60 ms) times the number of measurements to average. Assuming the sensor itself times out after approximately 24 ms, this means we cannot distinguish a sensor timeout from an object that is actually 4 m away.

Basically, it's up to the user to determine if the result is realistic.

File: pslab/external/hcsr04.py (hunk 1 above)
Comment (CloudyPadmal) [reply pslab/external/hcsr04.py:23]: Looks like the sensor I'm using (HC-SR04P) keeps the echo pin high without any internal timeout. 

When I tried the code snippet from [firmware-pr](https://github.com/fossasia/pslab-firmware/pull/118#issuecomment-1076290022), the LA capture list always miss the last entry (could be that it's missing the last falling edge).

File: pslab/external/hcsr04.py [hunk 3]
```diff
@@ -0,0 +1,118 @@
+"""Ultrasonic distance sensors."""
```
Comment (CloudyPadmal) [inline pslab/external/hcsr04.py:1]: Should we have the file name in Uppercase?

File: pslab/external/hcsr04.py [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    Example
+    -------
+    In this example the sensor's Vcc pin is connected to the PSLab's PV1 pin,
+    the Trig pin is connected to SQ1, Echo to LA1, and Gnd to GND.
+
+    >>> import pslab
+    >>> from pslab.external.hcsr04 import HCSR04
+    >>> psl = pslab.ScienceLab()
+    >>> distance_sensor = HCSR04(psl)
+    >>> psl.power_supply.pv1 = 5  # Power the sensor.
```
Comment (CloudyPadmal) [inline pslab/external/hcsr04.py:51]: One can also use `VCC` pin as the 5V source.

**Review by CloudyPadmal — COMMENTED** [review]
I tried to run the following code block

```python
import pslab; 
from pslab.external.hcsr04 import HCSR04
psl = pslab.ScienceLab()
distance_sensor = HCSR04(psl)
distance_sensor.estimate_distance()
```

but it's throwing some strange errors.

```bash
Traceback (most recent call last):
  File "/Users/padmal/Desktop/FOSSASIA/VENV/test_sonar.py", line 123, in <module>
    distance_sensor.estimate_distance()
  File "/Users/padmal/Desktop/FOSSASIA/VENV/.venv/lib/python3.7/site-packages/pslab/external/hcsr04.py", line 99, in estimate_distance
    block=False,
  File "/Users/padmal/Desktop/FOSSASIA/VENV/.venv/lib/python3.7/site-packages/pslab/instrument/logic_analyzer.py", line 322, in capture
    c = self._channels[c]
KeyError: 'LA1'
```
I tried printing out the `self._channels` list and instead of digital inputs, it's printing digital outputs.
```
{'SQ1': <pslab.instrument.digital.DigitalOutput object at 0x7fe000aa6f90>, 'SQ2': <pslab.instrument.digital.DigitalOutput object at 0x7fe000aa6fd0>, 'SQ3': <pslab.instrument.digital.DigitalOutput object at 0x7fe000aae050>, 'SQ4': <pslab.instrument.digital.DigitalOutput object at 0x7fe000aae090>}
```

It also causes a timeout for ACK error when lighting up the RGB LED. I had to comment it out to get it run.

<hr>

### I wonder if I'm running this right.
I created a virtual env and pulled this PR and used `pip install .` to setup the project.

<hr>

The snippet you shared in firmware PR works but I changed it a bit to put into a script as follows.

```python
import time
from pslab import LogicAnalyzer
from pslab import PWMGenerator
pw = PWMGenerator()
la = LogicAnalyzer()

_measure_period = 60e-3
_trigger_pulse_length = 10e-6

def estimate_distance(
        trig="SQ1",
        echo="LA1",
        average=10,
        speed_of_sound=340
    ):
    la.capture(channels=echo, events=2*average, block=False)
    pw.generate(
        channels=trig,
        frequency=_measure_period ** -1,
        duty_cycles=_trigger_pulse_length / _measure_period
    )
    time.sleep(_measure_period * average)
    pw.set_state(**{trig.lower(): 0})
    (t,) = la.fetch_data()

    high_times = -t[::2] + t[1::2]
    return speed_of_sound * high_times.mean() / 2 * 1e-6

print(estimate_distance()*100)
```

File: pslab/external/hcsr04.py (hunk 3 above)
Comment (bessman) [reply pslab/external/hcsr04.py:1]: Python module names should be lowercase. https://peps.python.org/pep-0008/#package-and-module-names

File: pslab/external/hcsr04.py (hunk 4 above)
Comment (bessman) [reply pslab/external/hcsr04.py:51]: Yes, or another SQ* pin set high. I prefer to leave the details of how to power the sensor to the user.

**Comment by bessman** [comment]
> I tried printing out the `self._channels` list and instead of digital inputs, it's printing digital outputs.

Yeah, I see why now. I was trying to be clever by having the `HCSR04` class inherit from `LogicAnalyzer`
 and `PWMGenerator` instead of uses instances of them. But the instruments are not designed for cooperative multiple inheritance, so the `PWMGenerator._channels` attribute overwrote `LogicAnalyzer._channels` since it's later in the MRO.

I want to fix the inheritance model of pslab-python so that derived instruments and sensors can use cooperative multiple inheritance, but for now I'll fix it by adding instrument instances to the sensor class instead.

I'm surprised that didn't happen earlier when you ran the code snippet in the other PR.

> It also causes a timeout for ACK error when lighting up the RGB LED. I had to comment it out to get it run.

This is because [8c0e975](https://github.com/fossasia/pslab-firmware/commit/8c0e9758a6d6de61b12792369f39f6bb5904d460) removed the SET_RGB1/2/3 subcommand and replaced them with SET_RGB_COMMON. I commented out that part of sciencelab.py a long time ago and forgot about it. I'll open a PR to fix it.

> The snippet you shared in firmware PR works but I changed it a bit to put into a script as follows.

Yep, those changes explain why it's working. They circumvent the inheritance problems outlined above.

File: pslab/external/hcsr04.py (hunk 1 above)
Comment (bessman) [reply pslab/external/hcsr04.py:23]: Does the ECHO pin eventually go back to LOW? IF so, does the method capture the final edge if you give it more time, say an extra 60 ms or so?

Comment (CloudyPadmal) [reply pslab/external/hcsr04.py:23]: I will check and confirm

File: pslab/external/hcsr04.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+            channels=self._trig,
+            frequency=self._measure_period**-1,
+            duty_cycles=self._trigger_pulse_length / self._measure_period,
+        )
+        # Wait one extra period to make sure we don't miss the final edge.
+        time.sleep(self._measure_period * (average + 1))
+        self._pwm.set_state(**{self._trig.lower(): 0})
+        (t,) = self._la.fetch_data()
+        self._sanity_check(len(t), 2 * average)
+        high_times = t[::2] - t[1::2]
```
Comment (CloudyPadmal) [inline pslab/external/hcsr04.py:111]: This should be other way around
```suggestion
        high_times = t[1::2] - t[::2]
```

**Review by CloudyPadmal — CHANGES_REQUESTED** [review]

File: pslab/external/hcsr04.py (hunk 1 above)
Comment (CloudyPadmal) [reply pslab/external/hcsr04.py:23]: Hi Alex, so I checked the ECHO pin status using an oscilloscope. Looks like it eventually goes to LOW after 200 ms.

<table>
<tr>
<td></td>
<td></td>
</tr>
</table>

This causes it to throw the following error (when it is a timeout)

```
<script_path>: RuntimeWarning: Mean of empty slice.
  return speed_of_sound * high_times.mean() / 2 * 1e-6
python3.7/site-packages/numpy/core/_methods.py:189: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
nan
```

**Review by CloudyPadmal — APPROVED** [review]
This implementation is working fine when the sensor is in range. It might need some error handling parts but that's not very critical. So I'm approving this PR.

**Pull request merged**